core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionImpl.java

```diff
@@ -16,60 +16,76 @@
-import java.util.Arrays;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.stream.Collectors;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
 
 import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
 import com.linecorp.armeria.server.ServiceRequestContext;
 
 /**
- * A {@link DynamicHttpFunction} implementation backed by Java Reflection.
+ * A {@link DynamicHttpFunction} implementation backed by Java Reflection. Methods whose
+ * return type is not {@link CompletionStage} or {@link HttpResponse} will be run in the
+ * blocking task executor.
  *
  * @see DynamicHttpServiceBuilder#addMappings(Object)
  */
 final class DynamicHttpFunctionImpl implements DynamicHttpFunction {
 
     private final Object object;
     private final Method method;
-    private final ParameterEntry[] parameterEntries;
+    private final List<ParameterEntry> parameterEntries;
+    private final boolean isSynchronous;
 
     DynamicHttpFunctionImpl(Object object, Method method) {
         this.object = requireNonNull(object, "object");
         this.method = requireNonNull(method, "method");
         this.parameterEntries = Methods.parameterEntries(method);
+        this.isSynchronous = !(
+                HttpResponse.class.isAssignableFrom(method.getReturnType()) ||
+                CompletionStage.class.isAssignableFrom(method.getReturnType()));
     }
 
     /**
      * Returns the set of parameter names, which should be provided to invoke this function.
      */
     Set<String> parameterNames() {
-        return Arrays.stream(parameterEntries).map(ParameterEntry::getName).collect(Collectors.toSet());
+        return parameterEntries.stream().map(ParameterEntry::getName).collect(toImmutableSet());
     }
 
     /**
      * Returns array of parameters for method invocation.
      */
     private Object[] parameters(Map<String, String> args) {
-        Object[] parameters = new Object[parameterEntries.length];
-        for (int i = 0; i < parameterEntries.length; ++i) {
-            ParameterEntry entry = parameterEntries[i];
-            String variable = entry.getName();
-            String value = args.get(variable);
-            assert value != null;
-            Class<?> type = entry.getType();
-            parameters[i] = Deserializers.deserialize(value, type);
-        }
-        return parameters;
+        return parameterEntries.stream()
+                               .map(entry -> {
+                                   String variable = entry.getName();
+                                   String value = args.get(variable);
+                                   assert value != null;
+                                   Class<?> type = entry.getType();
+                                   return Deserializers.deserialize(value, type);
+                               })
+                               .toArray(Object[]::new);
```

Comment:
Switched to ```toArray``` since the return type was Object.

Stream.toArray preallocates the result when the size is known, which it usually is. I guess it's a good replacement of the self-invoking weird syntax of List.toArray :)